### PR TITLE
Feature/permissions and design

### PR DESCRIPTION
**Pull request by ricdezen**
## Premise

Even though these are 1600 lines. There are 800 deletions because a lot of this was copy and pasting. Also most of the new code is xml.

## Changes

- Turned all Activities into Fragments, except for `AlarmAndLocateResponseActivity`
- Changed the UI into something that I personally liked, feel free to suggest changes.
- Added the [Android navigation component](https://developer.android.com/guide/navigation) implementation to make Fragment managing very easy. Basically there is the navigation graph xml file under the navigation directory, which contains a visual representation on how the app can be navigated from inside the various fragments, plus allows binding destinations to a menu item.
- The only Toolbar is the one in the `NavHolderActivity`, its title already changes based on the current fragment.
- The only notable name change in the activity names is `MainActivity` which is now `HomeFragment`, the rest got only the "Activity" suffix changed.
- Added a class for managing the permissions. Added dialogs to show if the user refuses the permissions.

## Side Changes

- Some minor changes to old code while moving things around.
- `LogItemFormatter` now searches the Contact database to know whether an Address matches a Contact, so `SMSLogEvent` now keeps only the address of the contact.
- Minor fixes to some tests, mostly due to the above change, so I added the `TestUtils` class containing some static methods to reset our various singletons to avoid the infamous SQL pointer error.

### Known issues

If you tap the back arrow/back button in the Log/Settings after accessing it from some other screen that's not the home, it will go to the home. I can probably fix this.

### Footer

Sorry if it took me a lot but I just kept seing things that needed to be fixed and couldn't help but try and fix them. I recommend downloading the branch and trying to actually run the app to see the result.
Also, apparently there has been a [covid-19 case in my town](https://www.youtube.com/watch?v=dQw4w9WgXcQ) , it's been a pleasure knowing y'all.

### CODE REVIEW

**Comment by RicDeZen** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 6
           

Complexity increasing per file
==============================
- app/src/main/java/ingsw/group1/findmyphone/fragment/ContactListFragment.java  2
- app/src/main/java/ingsw/group1/findmyphone/log/LogManager.java  1
- app/src/main/java/ingsw/group1/findmyphone/fragment/CreateContactFragment.java  3
- app/src/main/java/ingsw/group1/findmyphone/activity/NavHolderActivity.java  4
- app/src/test/java/ingsw/group1/findmyphone/RandomSMSLogEventGeneratorTest.java  1
- app/src/androidTest/java/ingsw/group1/findmyphone/contacts/SMSContactManagerTest.java  1
- app/src/main/java/ingsw/group1/findmyphone/PermissionInfoDialog.java  3
- app/src/main/java/ingsw/group1/findmyphone/PermissionHelper.java  5
- app/src/test/java/ingsw/group1/findmyphone/TestUtils.java  6
- app/src/main/java/ingsw/group1/findmyphone/fragment/ModifyContactFragment.java  2
- app/src/main/java/ingsw/group1/findmyphone/fragment/HomeFragment.java  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/RicDeZen/ingsw-group1-findmyphone/pullRequest?prid=5082269&bid=16717355)

File: app/src/main/java/ingsw/group1/findmyphone/PermissionInfoDialog.java [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+     * Default constructor. Takes the type of Dialog. Only accepted values are
+     * {@link PermissionInfoDialog#LOCATION} and {@link PermissionInfoDialog#MESSAGES}. Other
+     * values
+     * will result in incorrect or empty dialogs.
+     *
+     * @param type The type of dialog. See public constants of this class.
+     */
+    public PermissionInfoDialog(int type) {
+        super();
+        switch (type) {
```
Comment (RicDeZen) [inline app/src/main/java/ingsw/group1/findmyphone/PermissionInfoDialog.java]: Codacy Issue found: [A switch statement does not contain a break](https://app.codacy.com/manual/RicDeZen/ingsw-group1-findmyphone/pullRequest?prid=5082269)

File: app/src/main/java/ingsw/group1/findmyphone/fragment/HomeFragment.java [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+
+/**
+ * @author Turcato, Kumar
+ */
+public class HomeFragment extends Fragment {
+
+    private static final String INVALID_ADDRESS_MESSAGE = "The provided phone address is invalid";
+
+    private EditText addressInput;
+    private EditText passwordInput;
```
Comment (RicDeZen) [inline app/src/main/java/ingsw/group1/findmyphone/fragment/HomeFragment.java]: Codacy Issue found: [Perhaps 'passwordInput' could be replaced by a local variable.](https://app.codacy.com/manual/RicDeZen/ingsw-group1-findmyphone/pullRequest?prid=5082269)

File: app/src/main/java/ingsw/group1/findmyphone/PermissionHelper.java [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+            Manifest.permission.READ_SMS,
+            Manifest.permission.SEND_SMS,
+            Manifest.permission.RECEIVE_SMS,
+            Manifest.permission.READ_PHONE_STATE,
+    };
+
+    /**
+     * Permission required to bind a notification listener service.
+     */
+    private static final String[] NOTIF_SERVICE_PERMISSION =
```
Comment (RicDeZen) [inline app/src/main/java/ingsw/group1/findmyphone/PermissionHelper.java]: Codacy Issue found: [Avoid unused private fields such as 'NOTIF_SERVICE_PERMISSION'.](https://app.codacy.com/manual/RicDeZen/ingsw-group1-findmyphone/pullRequest?prid=5082269)

File: app/src/main/java/ingsw/group1/findmyphone/fragment/HomeFragment.java [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+                    manager.sendLocationRequest(smsPeer);
+                else showToastIfPossible(INVALID_ADDRESS_MESSAGE);
+            }
+        });
+
+        sendAlarmRequestButton.setOnClickListener(new View.OnClickListener() {
+            @Override
+            public void onClick(View v) {
+                smsPeer = new SMSPeer(addressInput.getText().toString());
+                String password = addressInput.getText().toString();
```
Comment (RicDeZen) [inline app/src/main/java/ingsw/group1/findmyphone/fragment/HomeFragment.java]: Codacy Issue found: [Avoid unused local variables such as 'password'.](https://app.codacy.com/manual/RicDeZen/ingsw-group1-findmyphone/pullRequest?prid=5082269)

File: app/src/main/java/ingsw/group1/findmyphone/fragment/HomeFragment.java (hunk 2 above)
Comment (RicDeZen) [inline app/src/main/java/ingsw/group1/findmyphone/fragment/HomeFragment.java]: Codacy Issue found: [Avoid unused private fields such as 'passwordInput'.](https://app.codacy.com/manual/RicDeZen/ingsw-group1-findmyphone/pullRequest?prid=5082269)

File: app/src/test/java/ingsw/group1/findmyphone/log/LogManagerTest.java [hunk 5]
```diff
@@ -26,8 +33,10 @@
 @RunWith(RobolectricTestRunner.class)
 public class LogManagerTest {
 
+    private static final String DB_NAME = LogManager.DEFAULT_LOG_DATABASE;
     private static final int TEST_LOG_SIZE = 100;
 
+    private SMSContactManager contacts;
```
Comment (RicDeZen) [inline app/src/test/java/ingsw/group1/findmyphone/log/LogManagerTest.java]: Codacy Issue found: [Perhaps 'contacts' could be replaced by a local variable.](https://app.codacy.com/manual/RicDeZen/ingsw-group1-findmyphone/pullRequest?prid=5082269)

File: app/src/main/java/ingsw/group1/findmyphone/fragment/HomeFragment.java [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+
+        if (getContext() == null) return root;
+
+        manager = new Manager(getContext());
+
+        sendLocationRequestButton.setOnClickListener(new View.OnClickListener() {
+            @Override
+            public void onClick(View v) {
+                smsPeer = new SMSPeer(addressInput.getText().toString());
+                String password = addressInput.getText().toString();
```
Comment (kummy98) [inline app/src/main/java/ingsw/group1/findmyphone/fragment/HomeFragment.java]: I think that was passwordInput.getText ....

**Comment by turcato-niccolo** [comment]
Some tests failed on my machine, I'm sorry today I can't look them out
image

```
[Robolectric] ingsw.group1.findmyphone.log.LogViewHolderTest.holderDoesNotExpandNonExpandable: sdk=28; resources=BINARY

android.view.InflateException: Binary XML file line #2: Binary XML file line #2: Error inflating class androidx.constraintlayout.widget.ConstraintLayout

Caused by: android.view.InflateException: Binary XML file line #2: Error inflating class androidx.constraintlayout.widget.ConstraintLayout
Caused by: java.lang.reflect.InvocationTargetException
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
	at android.view.LayoutInflater.$$robo$$android_view_LayoutInflater$createView(LayoutInflater.java:647)
	at android.view.LayoutInflater.createView(LayoutInflater.java)
	at android.view.LayoutInflater.$$robo$$android_view_LayoutInflater$createViewFromTag(LayoutInflater.java:790)
	at android.view.LayoutInflater.createViewFromTag(LayoutInflater.java)
	at android.view.LayoutInflater.$$robo$$android_view_LayoutInflater$createViewFromTag(LayoutInflater.java:730)
	at android.view.LayoutInflater.createViewFromTag(LayoutInflater.java)
	at android.view.LayoutInflater.$$robo$$android_view_LayoutInflater$inflate(LayoutInflater.java:492)
	at android.view.LayoutInflater.inflate(LayoutInflater.java)
	at android.view.LayoutInflater.$$robo$$android_view_LayoutInflater$inflate(LayoutInflater.java:423)
	at android.view.LayoutInflater.inflate(LayoutInflater.java)
	at android.view.LayoutInflater.$$robo$$android_view_LayoutInflater$inflate(LayoutInflater.java:374)
	at android.view.LayoutInflater.inflate(LayoutInflater.java)
	at ingsw.group1.findmyphone.log.LogViewHolderTest.createHolder(LogViewHolderTest.java:47)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.RunBefores.invokeMethod(RunBefores.java:33)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:24)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
```
```

[Robolectric] ingsw.group1.findmyphone.log.LogViewHolderTest.holderExpandsExpandable: sdk=28; resources=BINARY

android.view.InflateException: Binary XML file line #2: Binary XML file line #2: Error inflating class <unknown>

Caused by: android.view.InflateException: Binary XML file line #2: Error inflating class <unknown>
Caused by: java.lang.reflect.InvocationTargetException
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
	at android.view.LayoutInflater.$$robo$$android_view_LayoutInflater$createView(LayoutInflater.java:647)
	at android.view.LayoutInflater.createView(LayoutInflater.java)
	at android.view.LayoutInflater.$$robo$$android_view_LayoutInflater$createViewFromTag(LayoutInflater.java:790)
	at android.view.LayoutInflater.createViewFromTag(LayoutInflater.java)
	at android.view.LayoutInflater.$$robo$$android_view_LayoutInflater$createViewFromTag(LayoutInflater.java:730)
	at android.view.LayoutInflater.createViewFromTag(LayoutInflater.java)
	at android.view.LayoutInflater.$$robo$$android_view_LayoutInflater$inflate(LayoutInflater.java:492)
	at android.view.LayoutInflater.inflate(LayoutInflater.java)
	at android.view.LayoutInflater.$$robo$$android_view_LayoutInflater$inflate(LayoutInflater.java:423)
	at android.view.LayoutInflater.inflate(LayoutInflater.java)
	at android.view.LayoutInflater.$$robo$$android_view_LayoutInflater$inflate(LayoutInflater.java:374)
	at android.view.LayoutInflater.inflate(LayoutInflater.java)
	at ingsw.group1.findmyphone.log.LogViewHolderTest.createHolder(LogViewHolderTest.java:47)
```

**Comment by ricdezen** [comment]
That's very weird, I made sure all of them passed, I'll have to check again when I get home

**Comment by ricdezen** [comment]
Ok I confirm that they do pass on my pc, I can only assume it's due to something in the robolectric api glitching out, I'll have to dig around and find the problem.

**Comment by RicDeZen** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 4
           

Complexity increasing per file
==============================
- app/src/test/java/ingsw/group1/findmyphone/TestUtils.java  6
- app/src/main/java/ingsw/group1/findmyphone/fragment/ModifyContactFragment.java  2
- app/src/main/java/ingsw/group1/findmyphone/fragment/HomeFragment.java  2
- app/src/main/java/ingsw/group1/findmyphone/fragment/CreateContactFragment.java  3
- app/src/main/java/ingsw/group1/findmyphone/activity/NavHolderActivity.java  4
- app/src/androidTest/java/ingsw/group1/findmyphone/contacts/SMSContactManagerTest.java  1
- app/src/main/java/ingsw/group1/findmyphone/fragment/ContactListFragment.java  2
- app/src/main/java/ingsw/group1/findmyphone/PermissionInfoDialog.java  3
- app/src/test/java/ingsw/group1/findmyphone/RandomSMSLogEventGeneratorTest.java  1
- app/src/main/java/ingsw/group1/findmyphone/PermissionHelper.java  5
- app/src/main/java/ingsw/group1/findmyphone/log/LogManager.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/RicDeZen/ingsw-group1-findmyphone/pullRequest?prid=5082269&bid=16717355)

File: app/src/main/java/ingsw/group1/findmyphone/PermissionHelper.java (hunk 3 above)
Comment (RicDeZen) [inline app/src/main/java/ingsw/group1/findmyphone/PermissionHelper.java]: Codacy Issue found: [Avoid unused private fields such as 'NOTIF_SERVICE_PERMISSION'.](https://app.codacy.com/manual/RicDeZen/ingsw-group1-findmyphone/pullRequest?prid=5082269)

File: app/src/main/java/ingsw/group1/findmyphone/PermissionInfoDialog.java (hunk 1 above)
Comment (RicDeZen) [inline app/src/main/java/ingsw/group1/findmyphone/PermissionInfoDialog.java]: Codacy Issue found: [A switch statement does not contain a break](https://app.codacy.com/manual/RicDeZen/ingsw-group1-findmyphone/pullRequest?prid=5082269)

File: app/src/test/java/ingsw/group1/findmyphone/log/LogManagerTest.java [hunk 7]
```diff
@@ -26,8 +33,10 @@
 @RunWith(RobolectricTestRunner.class)
 public class LogManagerTest {
 
-    private static final int TEST_LOG_SIZE = 100;
+    private static final String DB_NAME = LogManager.DEFAULT_LOG_DATABASE;
+    private static final int TEST_LOG_SIZE = 10;
 
+    private SMSContactManager contacts;
```
Comment (RicDeZen) [inline app/src/test/java/ingsw/group1/findmyphone/log/LogManagerTest.java]: Codacy Issue found: [Perhaps 'contacts' could be replaced by a local variable.](https://app.codacy.com/manual/RicDeZen/ingsw-group1-findmyphone/pullRequest?prid=5082269)

File: app/src/main/java/ingsw/group1/findmyphone/fragment/HomeFragment.java [hunk 8]
```diff
@@ -0,0 +1,110 @@
...
+                    manager.sendLocationRequest(smsPeer);
+                else showToastIfPossible(INVALID_ADDRESS_MESSAGE);
+            }
+        });
+
+        sendAlarmRequestButton.setOnClickListener(new View.OnClickListener() {
+            @Override
+            public void onClick(View v) {
+                smsPeer = new SMSPeer(addressInput.getText().toString());
+                String password = passwordInput.getText().toString();
```
Comment (RicDeZen) [inline app/src/main/java/ingsw/group1/findmyphone/fragment/HomeFragment.java]: Codacy Issue found: [Avoid unused local variables such as 'password'.](https://app.codacy.com/manual/RicDeZen/ingsw-group1-findmyphone/pullRequest?prid=5082269)

**Comment by bortoletti-giorgia** [comment]
In my computer all tests pass.
I really like the graphics, the use of Fragments and also the management of Logs. Great job as always.

I don't know if I did something wrong but is it normal that after a request, it is not saved in the Log automatically? Maybe because the requests are saved after a response from the other side (?)

**Comment by turcato-niccolo** [comment]
I think the log is not wired in

Ok then it must be my android studio that has gone crazy, it was not wise to update to 3.6

**Comment by ricdezen** [comment]
The Log, like the encryption is not wired in yes. Btw I did not check the location request, something is probably wrong in the service

**Comment by ricdezen** [comment]
> it was not wise to update to 3.6

I did update and everything seems fine to me

**Comment by bortoletti-giorgia** [comment]
> The Log, like the encryption is not wired in yes. Btw I did not check the location request, something is probably wrong in the service

For the position something is wrong and in fact I opened an Issue.
However everything is ok and I approve the PR.

**Review by bortoletti-giorgia — DISMISSED** [review]

**Comment by ricdezen** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 4
           

Complexity increasing per file
==============================
- app/src/main/java/ingsw/group1/findmyphone/PermissionHelper.java  5
- app/src/androidTest/java/ingsw/group1/findmyphone/contacts/SMSContactManagerTest.java  1
- app/src/main/java/ingsw/group1/findmyphone/fragment/ContactListFragment.java  2
- app/src/main/java/ingsw/group1/findmyphone/fragment/HomeFragment.java  2
- app/src/test/java/ingsw/group1/findmyphone/TestUtils.java  6
- app/src/main/java/ingsw/group1/findmyphone/activity/NavHolderActivity.java  4
- app/src/main/java/ingsw/group1/findmyphone/PermissionInfoDialog.java  3
- app/src/test/java/ingsw/group1/findmyphone/RandomSMSLogEventGeneratorTest.java  1
- app/src/main/java/ingsw/group1/findmyphone/log/LogManager.java  1
- app/src/main/java/ingsw/group1/findmyphone/fragment/ModifyContactFragment.java  2
- app/src/main/java/ingsw/group1/findmyphone/fragment/CreateContactFragment.java  3
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/RicDeZen/ingsw-group1-findmyphone/pullRequest?prid=5082269&bid=16717355)

File: app/src/test/java/ingsw/group1/findmyphone/log/LogManagerTest.java (hunk 7 above)
Comment (ricdezen) [inline app/src/test/java/ingsw/group1/findmyphone/log/LogManagerTest.java:39]: Codacy Issue found: [Perhaps 'contacts' could be replaced by a local variable.](https://app.codacy.com/manual/RicDeZen/ingsw-group1-findmyphone/pullRequest?prid=5082269)

File: app/src/main/java/ingsw/group1/findmyphone/fragment/HomeFragment.java (hunk 8 above)
Comment (ricdezen) [inline app/src/main/java/ingsw/group1/findmyphone/fragment/HomeFragment.java:83]: Codacy Issue found: [Avoid unused local variables such as 'password'.](https://app.codacy.com/manual/RicDeZen/ingsw-group1-findmyphone/pullRequest?prid=5082269)

File: app/src/main/java/ingsw/group1/findmyphone/PermissionHelper.java [hunk 9]
```diff
@@ -0,0 +1,176 @@
...
```
Comment (ricdezen) [inline app/src/main/java/ingsw/group1/findmyphone/PermissionHelper.java:47]: Codacy Issue found: [Avoid unused private fields such as 'NOTIF_SERVICE_PERMISSION'.](https://app.codacy.com/manual/RicDeZen/ingsw-group1-findmyphone/pullRequest?prid=5082269)

File: app/src/main/java/ingsw/group1/findmyphone/PermissionInfoDialog.java [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
```
Comment (ricdezen) [inline app/src/main/java/ingsw/group1/findmyphone/PermissionInfoDialog.java:45]: Codacy Issue found: [A switch statement does not contain a break](https://app.codacy.com/manual/RicDeZen/ingsw-group1-findmyphone/pullRequest?prid=5082269)

**Comment by ricdezen** [comment]
Prepare for trouble and make it double cause this is gonna throw conflicts around when merging

**Pull request merged**